### PR TITLE
CI: add bors.toml

### PR DESCRIPTION
**Pull request by kaspar030**
### Contribution description

This adds configuration for [bors](https://bors.tech).

This has been discussed in CI meetings. The intended workflow is:

1. regular PR commits get built by CI as usual, but doing a quick-build (building only a subset of all boards), unless "CI: full build" is set
2. once that passes and the PR is ACKed, instead of merging using the merge button, we use bors to merge by typing "bors merge".
3. bors takes the PR, pushes it to the "testing" branch
4. CI does a full build of "testing"
5. on success, bors merges testing into master

The main benefits are

1. a quickbuild / full build scheme ("regular" PR builds will be <5 minutes, but only fully built branches get merged into master)
2. bors does "rollups", meaning, if multiple PRs are to be "bors merged", bors merges them together, tests them together and merges them together. That's more efficient than doing this sequentially, and prevents semantic merge conflicts (e.g., PR a and b pass CI individually, but fail when both merged, breaking master).

Once fully implemented, using bors will be a workflow change (no more "pressing the green button", but typing "bors merge" instead). But we've been using bors for RIOT-OS/riotdocker for a long time now, the experience was good (as opposed to disruptive).

This PR just adds the necessary bors configuration, and once merged will allow using bors to merge PRs using "bors merge", but the previous workflow is still in place (manual merging can be done, and "regular" PR builds are still full builds).

### Issues/PRs references

bors tracking issue: #18911

### CODE REVIEW

**Comment by kaspar030** [comment]
@miri64 I think you have the best overview on the github actions, did I catch all PR checks in bors.toml?

**Comment by kaspar030** [comment]
> Not sure, which naming must be matched, but I guess with that you have the better experience.

Thanks for pointing that out, I think the action name is the name of the status, so it has to match how it looks in the branch protection settings. I've changed the PR accordingly.

File: bors.toml
Comment (miri64) [inline bors.toml:19]: Yes it is. When there is no review, but no special label set, the check-labels check will succeed :-).

Comment (miri64) [reply bors.toml:19]: (also it might be good to have it as a fallback in general)

Comment (kaspar030) [reply bors.toml:19]: my thoughts exactly. I'll update the comment.

**Comment by kaspar030** [comment]
Do we want the github actions to also run again on the bors "testing" branch? static-tests, tools-buildtest come to mind.

**Comment by miri64** [comment]
> Do we want the github actions to also run again on the bors "testing" branch? static-tests, tools-buildtest come to mind.

I'd would prefer that.

**Comment by kaspar030** [comment]
let's see if bors can finish building...

bors try

**Comment by kaspar030** [comment]
> Timed out.

yeah, bors needs to be taught that our CI takes more than *4* hours. :(

**Comment by miri64** [comment]
> > Timed out.
> 
> yeah, bors needs to be taught that our CI takes more than _4_ hours. :(

Well hopefully, with bors activated, it won't anymore ^^". Should we wait with the next `try` for a time, when the queue is less crowded?

**Comment by kaspar030** [comment]
> Well hopefully, with bors activated, it won't anymore ^^". Should we wait with the next `try` for a time, when the queue is less crowded?

Yeah. There's also the issue that murdock builds bors' "trying.tmp" branch because it interprets the list in .murdock.yml as regexes.

On the plus side, the github actions are now activated (through a commit in this branch) to also run on "staging" and "trying".

**Comment by miri64** [comment]
> Yeah. There's also the issue that murdock builds bors' "trying.tmp" branch because it interprets the list in .murdock.yml as regexes.

you mean [this](https://github.com/RIOT-OS/RIOT/blob/a6ff838a4b5bcfac099d5e9bdb941fd735871b4a/.murdock.yml#L15) list? What speaks against making it a proper YAML list (except visual bloat)?

```yaml
commit:
  skip_keywords: 
  - "ci_skip"
  - "murdock_skip"
```

**Comment by kaspar030** [comment]
Everything for using this is in place! marking ready for review.

**Comment by kaspar030** [comment]
bors try

**Comment by miri64** [comment]
> Ah, nope, you meant the list of branches. Nvm... ^^"

Maybe change the regex for the branch to `"trying$"`?

File: .murdock.yml
Comment (kaspar030) [inline .murdock.yml:15]: @aabadie murdock is building this despite `[ci skip]` being in the commit subject:

image

Comment (aabadie) [reply .murdock.yml:15]: can you try "ci_skip" (which is already defined) ?

Comment (kaspar030) [reply .murdock.yml:15]: I could on a random PR, that specific commit was generated by bors (which I don't control).

**Comment by kaspar030** [comment]
bors try

**Comment by kaspar030** [comment]
bors try

**Comment by kaspar030** [comment]
bors try-

**Comment by kaspar030** [comment]
bors try

**Comment by miri64** [comment]
From https://github.com/RIOT-OS/RIOT/issues/18911

> - [ ] change "master" branch protection to only require the "bors" status

There is no bors status in this PR now that bors trying has passed nor is there a status to select in the branch protection rules. How did we set it up in riotdocker?

**Comment by miri64** [comment]
Ah, apparently this only shows up, when the `merge` action is run.

**Comment by kaspar030** [comment]
With the release phase over, IMO it is time for this PR.
So the idea is to "bors merge". If bors manages to test & merge this by itself, it's configured correctly. So once this is ACKed, we can give it a try. Once merged, we'd disable regular merging.

File: .murdock.yml
Comment (kaspar030) [reply .murdock.yml:15]: I've added "ci_skip" to the topmost commit here: https://github.com/RIOT-OS/RIOT/pull/18586
It got correctly skipped!

**Comment by kaspar030** [comment]
bors try

**Comment by kaspar030** [comment]
bors cancel

**Review by maribu — APPROVED** [review]

**Comment by miri64** [comment]
bors merge ?

**Comment by kaspar030** [comment]
I think this is still blocked on the "ci skip" issue :(

**Comment by kaspar030** [comment]
bors merge

**Comment by miri64** [comment]
bors merge

**Comment by miri64** [comment]
> I think this is still blocked on the "ci skip" issue :(
> 
> _edit_ (we could still merge this or have it merged by bors, and not use it until that is sorted out)

Seems so :-/

grafik

what are those `.tmp` branches anyways?

**Comment by kaspar030** [comment]
> what are those `.tmp` branches anyways?

I think they're bors temporary branches. it uses ".tmp" to merge (multiple) PR into, then pushes the result to non-tmp. otherwise, CI might pick up multiple commits to non-tmp ...

Somehow Murdock-ng neither respects the strict branch regex nor the "ci skip" thing. I looked at the logs, they look confusing. @aabadie is busy this week...

**Comment by kaspar030** [comment]
bors cancel

**Comment by kaspar030** [comment]
(I stopped the murdock builds, let's try again when something changes)

**Comment by kaspar030** [comment]
bors ping

**Comment by kaspar030** [comment]
bors merge

**Comment by miri64** [comment]
bors status is still not set to success. Probably because the merge was to staging, not `master`?

**Comment by kaspar030** [comment]
> bors status is still not set to success. Probably because the merge was to staging, not `master`?

so the bors status of this PR will never go to success. but the staging branch "Murdock" status is not set (it is set to "queued" still) for the last bors build.

**Comment by kaspar030** [comment]
bors cancel

**Comment by kaspar030** [comment]
bors merge

**Comment by kaspar030** [comment]
I have the impression that one of the "staging" builds got lost in murdock somehow. let's try again.
(next thing that might fail is bors not having the right to merge to master)

**Comment by kaspar030** [comment]
Ah, we basically need to reduce the master branch protections to "bors" status, no pull request required and "bors may push".
I'll do so after lunch

**Comment by aabadie** [comment]
Is it possible to configure bors so it doesn't put the whole PR description in the merge commit message ?

**Comment by kaspar030** [comment]
previous branch protections for reference:

image

**Comment by kaspar030** [comment]
and the previous release branch settings:

image

**Comment by aabadie** [comment]
> What would be a good position? (we could add a marker to the github PR template).
> It's a bit annoying that the comments don't get filtered out. Maybe drop the initial comment, and add a marker to the template, maybe just before the testing instructions?

I would even use `### Contribution description` because it can also be quite long. The PR number is already referenced in the commit message and GitHub/Murdock turns it into a link to the PR anyway.

**Pull request merged**
